GA4

HardNorth opened the pull request.

HardNorth committed:
Update Statistics service on GA4

HardNorth committed:
Add interpreter version detection

HardNorth committed:
Avoid reference error

HardNorth committed:
Fix query parameters

HardNorth committed:
Small refactoring

HardNorth committed:
Test rename

HardNorth committed:
Add clientId module test

HardNorth committed:
Actions version update

HardNorth committed:
statistics.js tests and fixes

HardNorth committed:
jsdoc format fixes

HardNorth committed:
Test fixes

HardNorth committed:
Refactoring

HardNorth committed:
Small test fix

HardNorth committed:
Small fix

HardNorth committed:
Small fix

AmsterGet commented on statistics/statistics.js:
```diff
@@ -0,0 +1,44 @@
+      client_name: PJSON_NAME,
+      client_version: PJSON_VERSION,
+    };
+    if (agentParams && Object.hasOwn(agentParams, 'name') && agentParams.name) {
```
I suggest to use `targetObject.hasOwnProperty` instead as we declare the support of Node.js >=10.
`Object.hasOwn` supported only from [Node.js v16.9.0](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwn#browser_compatibility).
I plan to drop support of stale Node.js versions together with migration to TS.

HardNorth replied:
Done

AmsterGet commented on statistics/statistics.js:
```diff
@@ -0,0 +1,44 @@
+
+  async trackEvent() {
+    const requestBody = {
+      client_id: getClientId(),
```
I see that this property generated once for the particular client.
Is it to identify unique clients?

HardNorth replied:
Yes, by this ID we can see new / old users in statistics.

AmsterGet commented on statistics/clientId.js:
```diff
@@ -0,0 +1,37 @@
+// client-id.js
```
Please remove the comment if it is not required.

HardNorth replied:
Done

AmsterGet commented on statistics/constants.js:
```diff
@@ -0,0 +1,67 @@
+    /* eslint-enable */
+  }
+  return null;
+}
```
I'm not sure if we need to check and detect the browser version as this client is only designed to run in a Node.js environment.
Any attempt to request anything from the file system or use the native Node.js API will fail in the browser for this library.
In case of Cypress that can run the tests directly in browser, the reporter lives separately in the Node.js process.

HardNorth replied:
OK, removed

HardNorth committed:
Update on async code and add error handling

HardNorth committed:
Fix tests

HardNorth committed:
Fix tests

HardNorth committed:
Remove redundant test

HardNorth committed:
Review fixes

HardNorth committed:
Some code style updates

HardNorth committed:
Rename client ID module.

HardNorth commented:
@AmsterGet I found I missed that `trackEvent` function is async, so I updated everything in `client-id.js` to be async. Also I renamed `clientId.js` to `client-id.js` to be more consistent with other JS file names.

HardNorth committed:
Remove redundant newline
